java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java

```diff
@@ -78,6 +69,38 @@ public JDBCSink(JDBCSinkConfig config, TableSchema tableSchema) {
                             String.format(ERROR_REPORT_TEMPLATE, e.getSQLState(), e.getMessage()))
                     .asRuntimeException();
         }
+
+        // Upsert sink must have primary key
+        if (config.isUpsertSink() && pkColumnNames.isEmpty()) {
+            throw Status.FAILED_PRECONDITION
+                    .withDescription(
+                            "No primary key in downstream for table " + config.getTableName())
+                    .asRuntimeException();
+        }
+
+        try {
+            var upsertSql =
+                    jdbcDialect.getUpsertStatement(
+                            config.getTableName(),
+                            List.of(tableSchema.getColumnNames()),
+                            pkColumnNames);
+            // MySQL and Postgres have upsert SQL
+            assert (upsertSql.isPresent());
```

Comment:
Seems that assert is disabled in java by default. We may write our own assert method.

Reply:
We may also add `-ea` in the argline configuration of the surefire plugin.